fix(build-tools): Windows compatible clean policy

jason-ha opened the pull request:
single quotes `'` become part of argument on Windows
- policy: require double quotes `"` for globs
- fix resolver logic and new quote policy

add generalized argument validation helpers

jason-ha committed:
fix(build-tools): Windows compatible clean policy
rimraf treats single quotes `'` as path of glob on Windows
- change policy to require double quotes `"`
- fix resolver logic and add single to double support

tylerbutler commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -1097,31 +1097,31 @@ export const handlers: Handler[] = [
 					return;
 				}
-				json.scripts["clean"] = `${clean} ${missing.map((name) => `'${name}'`).join(" ")}`;
+				if (missing.length !== 0) {
```
nit: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v48.0.1/docs/rules/explicit-length-check.md

```suggestion
				if (missing.length > 0) {
```

jason-ha committed:
fix(build-tools): Windows compatible clean policy
Windows shell treats single quotes `'` as part of argument
- policy: require double quotes `"` for globs and none otherwise

+ add generalized argument validation helpers

jason-ha committed:
Update build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts

Co-authored-by: Tyler Butler <tyler@tylerbutler.com>

tylerbutler commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -367,6 +367,61 @@ function ensurePrivatePackagesComputed(): void {
+ * @returns array of ordered pairs of resolved `arg` strings (quotes and escapes resolved)
+ *  and `original` corresponding strings.
+ */
+function parseArgs(commandLine: string, { onlyDoubleQuotes }: { onlyDoubleQuotes: boolean }) {
```
I didn't realize you could destructure function arguments like this. Cool!

jason-ha replied:
Oh yes. Very good for boolean options to force naming them.

tylerbutler commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -1097,31 +1152,28 @@ export const handlers: Handler[] = [
-					return "'clean' script rimraf argument should have single quotes";
+			if (cleanScript) {
+				if (cleanScript !== getPreferredCommandLine(cleanScript)) {
+					return "'clean' script double quote the globs and only the globs";
```
nit:
```suggestion
					return "'clean' script should double quote the globs and only the globs";
```

jason-ha committed:
fix(build-tools): message correction
